sharding/shard.go

```diff
@@ -0,0 +1,231 @@
+)
+
+type shardBackend interface {
+	Get(k common.Hash) ([]byte, error)
+	Has(k common.Hash) bool
+	Put(k common.Hash, val []byte) error
+	Delete(k common.Hash) error
+}
+
+// Shard base struct.
+type Shard struct {
+	shardDB shardBackend
+	shardID *big.Int
+}
+
+// MakeShard creates an instance of a Shard struct given a shardID.
+func MakeShard(shardID *big.Int, shardDB shardBackend) *Shard {
+	return &Shard{
+		shardID: shardID,
+		shardDB: shardDB,
+	}
+}
+
+// ShardID gets the shard's identifier.
+func (s *Shard) ShardID() *big.Int {
+	return s.shardID
+}
+
+// ValidateShardID checks if header belongs to shard.
+func (s *Shard) ValidateShardID(h *CollationHeader) error {
+	if s.ShardID().Cmp(h.ShardID()) != 0 {
+		return fmt.Errorf("collation does not belong to shard %d but instead has shardID %d", h.ShardID().Int64(), s.ShardID().Int64())
+	}
+	return nil
+}
+
+// HeaderByHash looks up a collation header from the shardDB using the header's hash.
+func (s *Shard) HeaderByHash(hash *common.Hash) (*CollationHeader, error) {
+	encoded, err := s.shardDB.Get(*hash)
+	if err != nil {
+		return nil, fmt.Errorf("header not found: %v", err)
+	}
+
+	var header CollationHeader
+
+	stream := rlp.NewStream(bytes.NewReader(encoded), uint64(len(encoded)))
+	if err := header.DecodeRLP(stream); err != nil {
+		return nil, fmt.Errorf("could not decode RLP header: %v", err)
+	}
+
+	return &header, nil
+}
+
+// CollationByHash fetches full collation.
+func (s *Shard) CollationByHash(headerHash *common.Hash) (*Collation, error) {
+	header, err := s.HeaderByHash(headerHash)
+	if err != nil {
+		return nil, err
+	}
+	if header == nil {
+		return nil, fmt.Errorf("header not found")
+	}
+
+	body, err := s.BodyByChunkRoot(header.ChunkRoot())
+	if err != nil {
+		return nil, err
+	}
+	return &Collation{header: header, body: body}, nil
+}
+
+// CanonicalHeaderHash gets a collation header hash that has been set as
+// canonical for shardID/period pair
+func (s *Shard) CanonicalHeaderHash(shardID *big.Int, period *big.Int) (*common.Hash, error) {
+	key := canonicalCollationLookupKey(shardID, period)
+
+	// fetches the RLP encoded collation header corresponding to the key.
+	encoded, err := s.shardDB.Get(key)
+	if err != nil || len(encoded) == 0 {
+		return nil, fmt.Errorf("no canonical collation header set for period, shardID pair: %v", err)
+	}
+
+	// RLP decodes the header, computes its hash.
+	var header CollationHeader
+
+	stream := rlp.NewStream(bytes.NewReader(encoded), uint64(len(encoded)))
+	if err := header.DecodeRLP(stream); err != nil {
+		return nil, fmt.Errorf("could not decode RLP header: %v", err)
+	}
+
+	collationHash := header.Hash()
+	return &collationHash, nil
+}
+
+// CanonicalCollation fetches the collation set as canonical in the shardDB.
+func (s *Shard) CanonicalCollation(shardID *big.Int, period *big.Int) (*Collation, error) {
+	h, err := s.CanonicalHeaderHash(shardID, period)
+	if err != nil {
+		return nil, fmt.Errorf("hash not found: %v", err)
```

Comment:
This error message is misleading. please change to "error while getting canoncial header hash: %v" 

`h == nil` means the header was not found. Please add error for `h == nil`  since it could happen that the hash was not found and it seems to be required to get the collation

Reply:
if err != nil it means in this case that either the .Get from the shardDB failed or DecodeRLP failed. See:
```go
func (s *Shard) CanonicalHeaderHash(shardID *big.Int, period *big.Int) (*common.Hash, error) {
	key := canonicalCollationLookupKey(shardID, period)

	// fetches the RLP encoded collation header corresponding to the key.
	encoded, err := s.shardDB.Get(key)
	if err != nil || len(encoded) == 0 {
		return nil, fmt.Errorf("no canonical collation header set for period, shardID pair: %v", err)
	}

	// RLP decodes the header, computes its hash.
	var header CollationHeader

	stream := rlp.NewStream(bytes.NewReader(encoded), uint64(len(encoded)))
	if err := header.DecodeRLP(stream); err != nil {
		return nil, fmt.Errorf("could not decode RLP header: %v", err)
	}

	collationHash := header.Hash()
	return &collationHash, nil
}
```

so if header is only nil if there is an error